### PR TITLE
client basic 256sha256 / server sample / tests / various fixes

### PR DESCRIPTION
**Pull request by StalderT**
https://github.com/open62541/open62541/pull/1686 https://github.com/open62541/open62541/issues/1687

fix two memleak, there is still one left (same memleak in :  ```examples/encryption/client_basic128rsa15.c```)

```
[1601-01-01 00:00:00.000 (UTC+0000)] warn/userland	No CA trust-list provided. Any remote certificate will be accepted.
[1601-01-01 00:00:00.000 (UTC+0000)] info/network	TCP network layer listening on opc.tcp://debian:4840/
[1601-01-01 00:00:00.000 (UTC+0000)] info/network	Connection 7 | New connection over TCP from ::1
[1601-01-01 00:00:00.000 (UTC+0000)] info/channel	Creating a new SecureChannel
[1601-01-01 00:00:00.000 (UTC+0000)] warn/securitypolicy	No PKI plugin set. Accepting all certificates
[1601-01-01 00:00:00.000 (UTC+0000)] info/channel	Connection 7 | SecureChannel 1 | Opened SecureChannel
[1601-01-01 00:00:00.000 (UTC+0000)] info/channel	Connection 7 | SecureChannel 1 | CloseSecureChannel
[1601-01-01 00:00:00.000 (UTC+0000)] info/network	Connection 7 | Closed
[1601-01-01 00:00:00.000 (UTC+0000)] info/network	Connection 7 | New connection over TCP from ::1
[1601-01-01 00:00:00.000 (UTC+0000)] info/channel	Creating a new SecureChannel
[1601-01-01 00:00:00.000 (UTC+0000)] info/channel	Connection 7 | SecureChannel 2 | Opened SecureChannel
[1601-01-01 00:00:00.000 (UTC+0000)] info/session	Connection 7 | SecureChannel 2 | Session 027c6003-bbf3-cb07-eefe-efbee37e7777 | ActivateSession: Session activated
[1601-01-01 00:00:00.000 (UTC+0000)] info/session	Connection 7 | SecureChannel 2 | Session 027c6003-bbf3-cb07-eefe-efbee37e7777 | CloseSession
[1601-01-01 00:00:00.000 (UTC+0000)] info/channel	Connection 7 | SecureChannel 2 | CloseSecureChannel
[1601-01-01 00:00:00.000 (UTC+0000)] info/network	Connection 7 | Closed
[1601-01-01 00:00:00.000 (UTC+0000)] info/network	Shutting down the TCP network layer
100%: Checks: 1, Failures: 0, Errors: 0
==23865== 
==23865== HEAP SUMMARY:
==23865==     in use at exit: 16 bytes in 1 blocks
==23865==   total heap usage: 12,642 allocs, 12,641 frees, 5,275,142 bytes allocated
==23865== 
==23865== 16 bytes in 1 blocks are definitely lost in loss record 1 of 1
==23865==    at 0x4C2DBC5: calloc (vg_replace_malloc.c:711)
==23865==    by 0x11859E: Array_decodeBinary (ua_types_encoding_binary.c:514)
==23865==    by 0x1187B0: String_decodeBinary (ua_types_encoding_binary.c:555)
==23865==    by 0x11ABBD: decodeBinaryInternal (ua_types_encoding_binary.c:1477)
==23865==    by 0x11ACCC: UA_decodeBinary (ua_types_encoding_binary.c:1506)
==23865==    by 0x16A9FC: processServiceResponse (ua_client.c:402)
==23865==    by 0x11EED7: UA_SecureChannel_finalizeChunk (ua_securechannel.c:791)
==23865==    by 0x11F6C4: UA_SecureChannel_processChunk (ua_securechannel.c:1057)
==23865==    by 0x16AACC: client_processChunk (ua_client.c:427)
==23865==    by 0x11C94A: completeChunkTrampoline (ua_connection.c:192)
==23865==    by 0x11C83B: processChunk (ua_connection.c:144)
==23865==    by 0x11C8DF: UA_Connection_processChunks (ua_connection.c:169)
==23865== 
==23865== LEAK SUMMARY:
==23865==    definitely lost: 16 bytes in 1 blocks
==23865==    indirectly lost: 0 bytes in 0 blocks
==23865==      possibly lost: 0 bytes in 0 blocks
==23865==    still reachable: 0 bytes in 0 blocks
==23865==         suppressed: 0 bytes in 0 blocks
==23865== 
==23865== For counts of detected and suppressed errors, rerun with: -v
==23865== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
```

https://travis-ci.org/StalderT/open62541/jobs/358805573#L7676

@pro @jpfr can you review the PR and the directories structure

### CODE REVIEW

**Comment by StalderT** [comment]
I can not find how to solve the memleak. @jpfr @opcua-tsn-team-kalycito can you take a look?

**Comment by StalderT** [comment]
> I can not find how to solve the memleak.

Found and fixed

**Comment by StalderT** [comment]
@Pro Do you know how I can fix : https://travis-ci.org/open62541/open62541/jobs/358908008#L2034 ? Thanks in advance

**Comment by Pro** [comment]
Thanks for this contribution!

> @Pro Do you know how I can fix : https://travis-ci.org/open62541/open62541/jobs/358908008#L2034 ? 

The name is probably wrong, i.e. it is using two times `check_` prefix. Try to remove the prefix from the test definition here:
https://github.com/StalderT/open62541/blob/2ec410fe287e2d658f31799fea0f2f23799e524f/tests/CMakeLists.txt#L209
`add_test_valgrind(encryption_basic128rsa15  ...`

**Comment by StalderT** [comment]
Thanks. I will try.

**Comment by StalderT** [comment]
Do you plan to include client encryption (https://github.com/open62541/open62541/pull/1686) in 0.3 ? It's the best way to test the server part.

File: tools/travis/travis_osx_script.sh [hunk 1]
```diff
@@ -11,6 +11,8 @@ cmake -DCMAKE_BUILD_TYPE=Release -DUA_BUILD_EXAMPLES=ON -DUA_BUILD_SELFSIGNED_CE
 make selfsigned
 ls examples
 cp examples/server_cert.der ../
+cp ./examples/server_cert.der ../
```
Comment (Pro) [inline tools/travis/travis_osx_script.sh]: duplicate copy

File: tools/travis/travis_linux_script.sh [hunk 2]
```diff
@@ -97,6 +97,8 @@ else
     cp -r doc ../../
     cp -r doc_latex ../../
     cp ./examples/server_cert.der ../../
+    cp ./examples/server_cert.der ../
```
Comment (Pro) [inline tools/travis/travis_linux_script.sh]: Already copied one line above

**Comment by Pro** [comment]
Client encryption will most probably not be part of 0.3 since we already had a API freeze. At some point we have to do that, and unfortunately there will always be features, which are nearly ready, but would delay releases. 0.4 already has some really nice features like Pub/Sub, Events and Architectures, which are all close to be ready.
Client encryption is already at a very high quality, but non the less, it still needs some testing on a broader base before we can include it in the release.

@StalderT note that the sonarcloud travis issue will be fixed with #1693 (i.e. disabled on fork PRs). The fuzzer issue is related to relative paths: The example uses a relative path to load the certificate, and the fuzzer runs the examples in another directory. Maybe we can use a preprocessor define for the unit tests, which points to the project root. Like this we can set it accordingly for fuzzer and non-fuzzer tests.

File: examples/encryption/client_basic256sha256.c [hunk 3]
```diff
@@ -0,0 +1,186 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (Pro) [inline examples/encryption/client_basic256sha256.c]: @StalderT can you also change this to CC0 as mentioned in #1686

**Comment by StalderT** [comment]
> Client encryption will most probably not be part of 0.3 since we already had a API freeze. At some point we have to do that, and unfortunately there will always be features, which are nearly ready, but would delay releases. 0.4 already has some really nice features like Pub/Sub, Events and Architectures, which are all close to be ready.

I understand. But currently the encryption server part is not tested by travis / appveyor. The only / best way to do this is to have encryption in the client part. Including basic256sha256 and basic128rsa15 has significantly decrease the coverage. If we want to get 80% it may be difficult.

image

Client encryption does not change the API, it just adds a new function : ```UA_Client_secure_new```

**Comment by StalderT** [comment]
> The fuzzer issue is related to relative paths: The unit test uses a relative path to load the certificate, and the fuzzer runs the tests in another directory. Maybe we can use a preprocessor define for the unit tests, which points to the project root. Like this we can set it accordingly in CMake for fuzzer and non-fuzzer tests.

@Pro fixed but fuzzer is still not happy, but for another reason. Can you take a look? Thanks in advance.

**Comment by jpfr** [comment]
Both sides are valid.
Wanting to unit-test encryption and freezing the API.

Maybe we can make a first 0.3.1 service-release that includes some little additions.
Otherwise I agree with @Pro that we have to make the cut for 0.3 at some point.

Major features are slated for 0.4 of course.

**Comment by Pro** [comment]
I will have a look at the fuzzer issue later today.
Having a 0.3.1 release in the near future and trying to release 0.3 as soon as possible is probably the best option here, i.e. fixing all the PR and issues marked with 0.3 Milestone, release 0.3, and then create a 0.3.1 soon after which includes some API additions, but no API changes

**Comment by StalderT** [comment]
@Pro For information : 


With ```UA_BUILD_FUZZING_CORPUS```
```
export CC=clang-3.9
export CXX=clang++-3.9
cmake -DUA_BUILD_FUZZING_CORPUS=ON -DUA_BUILD_UNIT_TESTS=ON -DUA_ENABLE_ENCRYPTION=ON ..
./bin/tests/check_encryption_basic128rsa15

Running suite(s): Encryption
[1601-01-01 00:00:00.000 (UTC+0000)] warn/userland	No CA trust-list provided. Any remote certificate will be accepted.
[1601-01-01 00:00:00.000 (UTC+0000)] info/network	TCP network layer listening on opc.tcp://debian:4840/
[1601-01-01 00:00:00.000 (UTC+0000)] info/network	Connection 7 | New connection over TCP from ::1
[1601-01-01 00:00:00.000 (UTC+0000)] info/server	Dumping package /data/open62541/bko/corpus/encryption_basic128rsa15/00001_hel.bin
[1601-01-01 00:00:00.000 (UTC+0000)] info/server	Dumping package /data/open62541/bko/corpus/encryption_basic128rsa15/00002_opn.bin
[1601-01-01 00:00:00.000 (UTC+0000)] info/channel	Creating a new SecureChannel
[1601-01-01 00:00:00.000 (UTC+0000)] warn/securitypolicy	No PKI plugin set. Accepting all certificates
[1601-01-01 00:00:00.000 (UTC+0000)] info/channel	Connection 7 | SecureChannel 1 | Opened SecureChannel
[1601-01-01 00:00:00.000 (UTC+0000)] info/server	Dumping package /data/open62541/bko/corpus/encryption_basic128rsa15/00003_msg_GetEndpointsRequest.bin
[1601-01-01 00:00:00.000 (UTC+0000)] info/server	Dumping package /data/open62541/bko/corpus/encryption_basic128rsa15/00004_clo.bin
[1601-01-01 00:00:00.000 (UTC+0000)] info/channel	Connection 7 | SecureChannel 1 | CloseSecureChannel
[1601-01-01 00:00:00.000 (UTC+0000)] info/network	Connection 7 | Closed
[1601-01-01 00:00:00.000 (UTC+0000)] info/network	Connection 7 | New connection over TCP from ::1
[1601-01-01 00:00:00.000 (UTC+0000)] info/server	Dumping package /data/open62541/bko/corpus/encryption_basic128rsa15/00005_hel.bin
[1601-01-01 00:00:00.000 (UTC+0000)] info/server	Dumping package /data/open62541/bko/corpus/encryption_basic128rsa15/00006_opn.bin
[1601-01-01 00:00:00.000 (UTC+0000)] info/channel	Creating a new SecureChannel
[1601-01-01 00:00:00.000 (UTC+0000)] info/channel	Connection 7 | SecureChannel 2 | Opened SecureChannel
[1601-01-01 00:00:00.000 (UTC+0000)] info/server	Dumping package /data/open62541/bko/corpus/encryption_basic128rsa15/00007_msg_GetEndpointsRequest.bin
[1601-01-01 00:00:00.000 (UTC+0000)] info/network	Connection 7 | Processing the message failed with error BadSecurityChecksFailed
[1601-01-01 00:00:00.000 (UTC+0000)] info/network	Connection 7 | Closed
[1601-01-01 00:00:00.000 (UTC+0000)] error/client	GetEndpointRequest failed with error code BadSecureChannelIdInvalid
[1601-01-01 00:00:00.000 (UTC+0000)] info/network	Shutting down the TCP network layer
0%: Checks: 1, Failures: 1, Errors: 0
/data/open62541/tests/encryption/check_encryption_basic128rsa15.c:160:F:Encryption basic128rsa15:encryption_connect:0: Assertion 'retval == 0x00' failed: retval == 2149711872, 0x00 == 0
```

Without ```UA_BUILD_FUZZING_CORPUS```
```
export CC=clang-3.9
export CXX=clang++-3.9
cmake -DUA_BUILD_UNIT_TESTS=ON -DUA_ENABLE_ENCRYPTION=ON ..
./bin/tests/check_encryption_basic128rsa15

Running suite(s): Encryption
[1601-01-01 00:00:00.000 (UTC+0000)] warn/userland	No CA trust-list provided. Any remote certificate will be accepted.
[1601-01-01 00:00:00.000 (UTC+0000)] info/network	TCP network layer listening on opc.tcp://debian:4840/
[1601-01-01 00:00:00.000 (UTC+0000)] info/network	Connection 7 | New connection over TCP from ::1
[1601-01-01 00:00:00.000 (UTC+0000)] info/channel	Creating a new SecureChannel
[1601-01-01 00:00:00.000 (UTC+0000)] warn/securitypolicy	No PKI plugin set. Accepting all certificates
[1601-01-01 00:00:00.000 (UTC+0000)] info/channel	Connection 7 | SecureChannel 1 | Opened SecureChannel
[1601-01-01 00:00:00.000 (UTC+0000)] info/network	Connection 8 | New connection over TCP from ::1
[1601-01-01 00:00:00.000 (UTC+0000)] info/channel	Connection 7 | SecureChannel 1 | CloseSecureChannel
[1601-01-01 00:00:00.000 (UTC+0000)] info/network	Connection 7 | Closed
[1601-01-01 00:00:00.000 (UTC+0000)] info/channel	Creating a new SecureChannel
[1601-01-01 00:00:00.000 (UTC+0000)] info/channel	Connection 8 | SecureChannel 2 | Opened SecureChannel
[1601-01-01 00:00:00.000 (UTC+0000)] info/session	Connection 8 | SecureChannel 2 | Session 027c6003-bbf3-cb07-eefe-efbee37e7777 | ActivateSession: Session activated
[1601-01-01 00:00:00.000 (UTC+0000)] info/session	Connection 8 | SecureChannel 2 | Session 027c6003-bbf3-cb07-eefe-efbee37e7777 | CloseSession
[1601-01-01 00:00:00.000 (UTC+0000)] info/channel	Connection 8 | SecureChannel 2 | CloseSecureChannel
[1601-01-01 00:00:00.000 (UTC+0000)] info/network	Shutting down the TCP network layer
100%: Checks: 1, Failures: 0, Errors: 0
```
Removing ```UA_debug_dumpCompleteChunk``` in ```processCompleteChunk``` fix the issue
Removing call of ```UA_SecureChannel_processChunk``` in ```UA_debug_dumpCompleteChunk``` also fix the issue

**Comment by Pro** [comment]
> The problem probably comes because UA_SecureChannel_processChunk is called 2 times.

Yes, that's also what I found out in the afternoon. It is required to get the service name to create a useful file dump name. I will fix that somehow

**Comment by Pro** [comment]
@StalderT to fix the fuzz encryption backing up the message buffer won't help. The problem is that processChunk is not stateless, i.e. it stores some (encryption) data into the `connection->channel`. I need to find that and then we need to reset that value after the first processChunk call (as it is already done for `receiveSequenceNumber`)

**Comment by Pro** [comment]
Also note that you can check the fuzzer locally by running `tests/fuzz/generate_corpus.sh`

**Comment by StalderT** [comment]
> @StalderT to fix the fuzz encryption backing up the message buffer won't help. The problem is that processChunk is not stateless, i.e. it stores some (encryption) data into the connection->channel. I need to find that and then we need to reset that value after the first processChunk call (as it is already done for receiveSequenceNumber)

Ok. For the moment, with the incomplet fix, fuzzer tests pass.

When you are ready, PR can be merged in master.

https://github.com/open62541/open62541/pull/1691/commits/21920333db4f22bb4d797504819222d311e76217 must also be merged in 0.3

**Comment by Pro** [comment]
You were right with the messageBuffer. I changed it a bit in my last commit. 

The backport of the mentioned commit is in #1696 

Additionally I tried to disable Sonarcloud for PRs, let's see if all is green, then we merge.
Thanks @StalderT for your very nice contribution!!

**Comment by StalderT** [comment]
Should not be ```if ! [ "${TRAVIS_REPO_SLUG}" = "open62541/open62541" ] || (``` instead of ```if ! [ "${TRAVIS_REPO_SLUG}" = "open62541/open62541" ] && (``` ?

**Comment by StalderT** [comment]
@Pro sorry I made a mistake. Can you remove https://github.com/open62541/open62541/pull/1691/commits/8405c37dec273c4ac2df1e76e019d0f57cee9446 from the PR and resend your last ```Disable sonarcloud on PRs``` commit ?

Really sorry!!!

**Pull request merged**